### ISSUE TITLE
zip can only zip iterables of the same type

### ISSUE THREAD

**Issue by mreichelt** [leisim/dartx#91]
The zip method is especially useful when we can zip items of different types - but `zip` in dartx is implemented to only accept a right-hand iterable of the same type. So things like this work:

```dart
final numbers = [1, 2, 3];
final multipliers = [3, 4, 5];

final multiplied = numbers.zip(multipliers, (number, multiplier) => number * multiplier).toList();
print(multiplied); // prints [3, 8, 15]
```

But this doesn't:

```dart
final countOfFruits = [1, 2, 3];
final fruits = ['bananas', 'apples', 'oranges'];
final combined = count.zip(fruits, (count, fruit) => '$count $fruit').toList();
```

![image](https://user-images.githubusercontent.com/400322/86129185-9695db00-bae2-11ea-928f-61deb9410c01.png)

We can try to change the implementation of the original `zip`, but then we must make sure we don't break any calling code. So let's test this thoroughly. 🤓

**Comment by mreichelt** [comment]
Rough new implementation that works for me:

```dart
extension IterableZip<E> on Iterable<E> {
  Iterable<V> zipWith<R, V>(Iterable<R> other, V Function(E a, R b) transform) sync* {
    var it1 = iterator;
    var it2 = other.iterator;
    while (it1.moveNext() && it2.moveNext()) {
      yield transform(it1.current, it2.current);
    }
  }
}
```

**Comment by mreichelt** [comment]
Yep - this works, but unfortunately it also breaks consumer code:

![image](https://user-images.githubusercontent.com/400322/86498367-dcbe9900-bd85-11ea-99ce-6d637dee4cf5.png)

The second parameter will now become `dynamic` - which is sad, because every info for Dart to find out the generic `R` is available at build time. It just won't pass that as generic for parameter `b` in `transform`.
And Dart doesn't have operator overloading, so that doesn't work as well.

@leisim @passsy it would be great if you can add some guidance on how you would like this to be handled. I basically see two options:

1. Keep the existing implementation, and add the new one under a different name (i.e. `zipWith`). That works, but then we'll have two methods where one is more broad than the other. And we'll loose the nice naming, which I would like to stay similar with Kotlin. We could deprecate `zip`, but still we would not be able to keep the name.
2. Leverage the fact that dartx is still a 0.x version, and as such, can add breaking changes. Maybe we should ask the community first before we do this. I added a PR for this, just to show how this would look like and to start the discussion.

**Comment by leisim** [comment]
@mreichelt Thanks for the detailed proposal.

Is it only me or is this a bug in Dart? In my opinion Dart should be able to infer the types correct without manual annotations.

I don't think we should add a `zipWith()` method that has basically does the same as the `zip()` method. It is not great that type annotations will be necessary in the future but I think we should replace the existing `zip ()` method with the proposed one.

**Comment by mreichelt** [comment]
Ok, thanks! Then, at least I'll add some docs to the `zip` method later so people know how to make their code working again :)

**Comment by mreichelt** [comment]
@leisim I added an example in the `.zip()` documentation as stated. It's ready to be merged now IMO :-)
https://github.com/leisim/dartx/pull/94

**Comment by passsy** [comment]
That's a problem I've also ran into when building `kt.dart`. 

![Screen-Shot-2020-07-09-20-25-33 80](https://user-images.githubusercontent.com/1096485/87077442-6ad1ce00-c223-11ea-8bab-f355867502bd.png)

### Definitely a dart issue

I think right now would be a good time to open a dart-lang/sdk issue about this. I searched for quite a while but haven't found anything related.
<details>
<summary>details</summary>

First guesses:
It seems like Kotlin evaluates generic type parameters (`T`) in the arguments starting with the first argument. If types are provided, it will assume those types for all following arguments, also when an argument is a lambda of type `(T) -> Any` 
In Dart the generic types in the arguments don't seem to have precedence. The compiler doesn't forward previously defined generic types (`T`) to lambdas which are defined as `dynamic Function(dynamic)`.

TODO: come up with an easier example than `zip()` for the bug report

</details>


### Split zip in two methods
Since we're doing a breaking API change, I was thinking about aligning the `zip` functionality from kt_dart with dartx.

In kt_dart I've split zip into two methods
- `zip<R>(KtIterable<R> other): KtList<KtPair<T, R>>` [src](https://github.com/passsy/kt.dart/blob/28bba9f9c5b31525c1b91c889c4121dc51376b81/lib/src/collection/kt_iterable.dart#L1699-L1702)
Returns a pair with `T` and `R`
- `zipTransform<R, V>(KtIterable<R> other, V Function(T a, R b) transform): KtList<V>` [src](https://github.com/passsy/kt.dart/blob/28bba9f9c5b31525c1b91c889c4121dc51376b81/lib/src/collection/kt_iterable.dart#L1704-L1721)
Same as the proposed `zip` here.

I decided against it because Dart doesn't offer tuples in the Dart SDK. Therefore we are left with a single zip method and don't have to think about the naming conflict.

### Going forward

- [ ] Let's do the breaking change
- [ ] Create a Dart SDK bug report
- [ ] Document the new required type information adequately
- [ ] Wait for Dart SDK fix to make our documentation obsolete
